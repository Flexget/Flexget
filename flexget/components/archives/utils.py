"""
Utilities for handling RAR and ZIP archives

Provides wrapper archive and exception classes to simplify
archive extraction
"""


import os
import shutil
import zipfile

from loguru import logger

try:
    import rarfile
except ImportError:
    rarfile = None

logger = logger.bind(name='archive')


class ArchiveError(Exception):
    """Base exception for archive"""

    pass


class NeedRarFile(ArchiveError):
    """Exception to be raised when rarfile module is missing"""

    pass


class BadArchive(ArchiveError):
    """Wrapper exception for BadZipFile and BadRarFile"""

    pass


class NeedFirstVolume(ArchiveError):
    """Wrapper exception for rarfile.NeedFirstVolume"""

    pass


class PathError(ArchiveError):
    """Exception to be raised when an archive file doesn't exist"""

    pass


class FSError(ArchiveError):
    """Exception to be raised on OS/IO exceptions"""

    pass


class FileAlreadyExists(ArchiveError):
    """Exception to be raised when destination file already exists"""

    pass


def rarfile_set_tool_path(config):
    """
    Manually set the path of unrar executable if it can't be resolved from the
    PATH environment variable
    """
    unrar_tool = config['unrar_tool']

    if unrar_tool:
        if not rarfile:
            logger.error('rar_tool specified with no rarfile module installed.')
        else:
            rarfile.UNRAR_TOOL = unrar_tool
            logger.debug('Set RarFile.unrar_tool to: {}', unrar_tool)


def rarfile_set_path_sep(separator):
    """
    Set the path separator on rarfile module
    """
    if rarfile:
        rarfile.PATH_SEP = separator


def makepath(path):
    """Make directories as needed"""
    if not os.path.exists(path):
        logger.debug('Creating path: {}', path)
        os.makedirs(path)


class Archive:
    """
    Base archive class. Assumes an interface similar to
    zipfile.ZipFile or rarfile.RarFile
    """

    def __init__(self, archive_object, path):
        self.path = path

        self.archive = archive_object(self.path)

    def close(self):
        """Release open resources."""
        self.archive.close()

    def delete(self):
        """Delete the volumes that make up this archive"""
        volumes = self.volumes()
        self.close()

        try:
            for volume in volumes:
                os.remove(volume)
                logger.verbose('Deleted archive: {}', volume)
        except OSError as error:
            raise FSError(error)

    def volumes(self):
        """Returns the list of volumes that comprise this archive"""
        return [self.path]

    def infolist(self):
        """Returns a list of info objects describing the contents of this archive"""
        infolist = []

        for info in self.archive.infolist():
            try:
                archive_info = ArchiveInfo(info)
                infolist.append(archive_info)
            except ValueError as e:
                logger.debug(e)

        return infolist

    def open(self, member):
        """Returns file-like object from where the data of a member file can be read."""
        return self.archive.open(member)

    def extract_file(self, member, destination):
        """Extract a member file to the specified destination"""
        try:
            with self.open(member) as source:
                with open(destination, 'wb') as target:
                    shutil.copyfileobj(source, target)
        except OSError as error:
            raise FSError(error)


class RarArchive(Archive):
    """
    Wrapper class for rarfile.RarFile
    """

    def __init__(self, path):
        RarArchive.check_import()

        try:
            super().__init__(rarfile.RarFile, path)
        except rarfile.BadRarFile as error:
            raise BadArchive(error)
        except rarfile.NeedFirstVolume as error:
            raise NeedFirstVolume(error)
        except rarfile.Error as error:
            raise ArchiveError(error)

    def volumes(self):
        """Returns the list of volumes that comprise this archive"""
        return self.archive.volumelist()

    def open(self, member):
        """Returns file-like object from where the data of a member file can be read."""
        try:
            return super().open(member)
        except rarfile.Error as error:
            raise ArchiveError(error)

    @staticmethod
    def check_import():
        if not rarfile:
            raise NeedRarFile('Python module rarfile needed to handle RAR archives')


class ZipArchive(Archive):
    """
    Wrapper class for zipfile.ZipFile
    """

    def __init__(self, path):
        try:
            super().__init__(zipfile.ZipFile, path)
        except zipfile.BadZipfile as error:
            raise BadArchive(error)

    def open(self, member):
        """Returns file-like object from where the data of a member file can be read."""
        try:
            return super().open(member)
        except zipfile.BadZipfile as error:
            raise ArchiveError(error)


class ArchiveInfo:
    """Wrapper class for  archive info objects"""

    def __init__(self, info):
        self.info = info
        self.path = info.filename
        self.filename = os.path.basename(self.path)

        if self._is_dir():
            raise ValueError('Appears to be a directory: %s' % self.path)

    def _is_dir(self):
        """Indicates if info object looks to be a directory"""

        if hasattr(self.info, 'isdir'):
            return self.info.isdir()
        else:
            return not self.filename

    def extract(self, archive, destination):
        """Extract ArchiveInfo object to the specified destination"""
        dest_dir = os.path.dirname(destination)

        if os.path.exists(destination):
            raise FileAlreadyExists('File already exists: %s' % destination)

        logger.debug('Creating path: {}', dest_dir)
        makepath(dest_dir)

        try:
            archive.extract_file(self.info, destination)
            logger.verbose('Extracted: {} to {}', self.path, destination)
        except Exception as error:
            if os.path.exists(destination):
                logger.debug('Cleaning up partially extracted file: {}', destination)
                os.remove(destination)

            raise error


def open_archive(archive_path):
    """
    Returns the appropriate archive object
    """

    archive = None

    if not os.path.exists(archive_path):
        raise PathError('Path doesn\'t exist')

    if zipfile.is_zipfile(archive_path):
        archive = ZipArchive(archive_path)
        logger.debug('Successfully opened ZIP: {}', archive_path)
    elif rarfile and rarfile.is_rarfile(archive_path):
        archive = RarArchive(archive_path)
        logger.debug('Successfully opened RAR: {}', archive_path)
    else:
        if not rarfile:
            logger.warning('Rarfile module not installed; unable to handle RAR archives.')

    return archive


def is_archive(path):
    """
    Attempts to open an entry as an archive; returns True on success, False on failure.
    """

    archive = None

    try:
        archive = open_archive(path)
        if archive:
            archive.close()
            return True
    except (OSError, ArchiveError) as error:
        logger.debug('Failed to open file as archive: {} ({})', path, error)

    return False
