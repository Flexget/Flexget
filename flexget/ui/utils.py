def pretty_date(time=False):
    """
    Get a datetime object or a int() Epoch timestamp and return a
    pretty string like 'an hour ago', 'Yesterday', '3 months ago',
    'just now', etc
    """
    from datetime import datetime
    now = datetime.now()
    diff = now - datetime.fromtimestamp(time)
    second_diff = diff.seconds
    day_diff = diff.days

    if day_diff < 0:
        return ''

    if day_diff == 0:
        if second_diff < 10:
            return "just now"
        if second_diff < 60:
            return str(second_diff) + " seconds ago"
        if second_diff < 120:
            return "a minute ago"
        if second_diff < 3600:
            return str(second_diff / 60) + " minutes ago"
        if second_diff < 7200:
            return "an hour ago"
        if second_diff < 86400:
            return str(second_diff / 3600) + " hours ago"
    if day_diff == 1:
        return "Yesterday"
    if day_diff < 7:
        return str(day_diff) + " days ago"
    if day_diff < 31:
        return str(day_diff / 7) + " weeks ago"
    if day_diff < 365:
        return str(day_diff / 30) + " months ago"
    return str(day_diff / 365) + " years ago"
