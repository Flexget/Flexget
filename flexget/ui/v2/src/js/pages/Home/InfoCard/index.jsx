import React from 'react';
import { CardContent } from 'material-ui/Card';
import IconButton from 'material-ui/IconButton';
import Icon from 'material-ui/Icon';
import 'font-awesome/css/font-awesome.css';
import {
  InfoCardWrapper,
  InfoCardHeader,
  BoldParagraph,
  InfoCardActions,
} from './styles';

const InfoCard = () => (
  <InfoCardWrapper>
    <InfoCardHeader
      title="Flexget Web Interface"
      subheader="Under Development"
    />
    <CardContent>
      <BoldParagraph>
        We need your help! If you are a React developer or can help with the layout/design/css
        then please join in the effort!
      </BoldParagraph>
      <p>
        The interface is not yet ready for end users. Consider this preview only state.
      </p>
      <p>
        If you still use it anyways, please report back to us on how well it works, issues,
        ideas etc...
      </p>
      <p>
        There is a functional API with documentation available at <a href="/api">/api</a>
      </p>
      <p>
        More information: <a
          href="http://flexget.com/wiki/Web-UI/v2"
          target="_blank"
          rel="noopener noreferrer"
        >
          http://flexget.com/wiki/Web-UI/v2
        </a>
      </p>
      <p>
        Gitter Chat: <a
          href="https://gitter.im/Flexget/Flexget"
          target="_blank"
          rel="noopener noreferrer"
        >
          https://gitter.im/Flexget/Flexget
        </a>
      </p>
    </CardContent>
    <InfoCardActions>
      <IconButton
        aria-label="Github"
        href="https://github.com/Flexget/Flexget"
        target="_blank"
        rel="noopener noreferrer"
      >
        <Icon className="fa fa-github" />
      </IconButton>
      <IconButton
        aria-label="Flexget.com"
        href="https://flexget.com"
        target="_blank"
        rel="noopener noreferrer"
      >
        <Icon className="fa fa-home" />
      </IconButton>
      <IconButton
        aria-label="Gitter"
        href="https://gitter.im/Flexget/Flexget"
        target="_blank"
        rel="noopener noreferrer"
      >
        <Icon className="fa fa-comment" />
      </IconButton>
      <IconButton
        aria-label="Forum"
        href="https://discuss.flexget.com"
        target="_blank"
        rel="noopener noreferrer"
      >
        <Icon className="fa fa-forumbee" />
      </IconButton>
    </InfoCardActions>
  </InfoCardWrapper>
);

export default InfoCard;
