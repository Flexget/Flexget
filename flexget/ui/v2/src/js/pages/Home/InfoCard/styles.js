import styled from 'react-emotion';
import Card, { CardActions, CardHeader } from 'material-ui/Card';
import theme from 'theme';

export const InfoCardWrapper = styled(Card)`
  margin: 0 auto;
  ${theme.breakpoints.up('sm')} {
    width: 50%;
  }
`;

export const InfoCardHeader = styled(CardHeader)`
  background-color: ${theme.palette.primary[800]};
`;

export const BoldParagraph = styled.p`
  font-weight: bold;
`;

export const InfoCardActions = styled(CardActions)`
  justify-content: center;
`;
