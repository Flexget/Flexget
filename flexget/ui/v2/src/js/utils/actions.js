import { LOADING_STATUS } from 'store/status/actions';

export const action = (type, payload, {
  message,
} = {}) => ({
  type,
  payload,
  error: (payload instanceof Error),
  meta: {
    message,
  },
});

export const request = (type, payload = {}) => ({
  type: LOADING_STATUS,
  payload,
  meta: { type },
});

export const requesting = type => act => (
  act.type === LOADING_STATUS &&
    act.meta.type === type
);
