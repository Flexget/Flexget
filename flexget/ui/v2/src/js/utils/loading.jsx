import Loadable from 'react-loadable';
import Loading from 'common/Loader';

export function createAsyncComponent(loader) {
  return Loadable({
    loading: Loading,
    loader,
  });
}
