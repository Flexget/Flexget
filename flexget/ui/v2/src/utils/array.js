export function groupBy(array, key, transform = f => f) {
  return array.reduce((obj, curr) => {
    const modKey = transform(curr[key]);
    return {
      ...obj,
      [modKey]: obj[modKey] ? [...obj[modKey], curr] : [curr],
    };
  }, {});
}
