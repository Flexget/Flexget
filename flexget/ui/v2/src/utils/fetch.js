function status(response) {
  return response.json().then((data) => {
    if (response.status >= 200 && response.status < 300) {
      return { data, headers: response.headers };
    }
    const err = new Error(data.message);
    err.status = response.status;
    throw err;
  });
}

function request(resource, method, body) {
  const contentType = method === 'get' ? {} : { 'Content-Type': 'application/json' };
  return fetch(`/api${resource}`, {
    method,
    headers: {
      Accept: 'application/json',
      ...contentType,
    },
    credentials: 'same-origin',
    body: JSON.stringify(body),
  })
    .then(status);
}

export function get(resource) {
  return request(resource, 'get');
}

export function post(resource, body) {
  return request(resource, 'post', body);
}

export function put(resource, body) {
  return request(resource, 'put', body);
}

export function del(resource) {
  return request(resource, 'delete');
}
