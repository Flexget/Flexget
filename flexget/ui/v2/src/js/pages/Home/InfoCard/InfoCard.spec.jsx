import React from 'react';
import renderer from 'react-test-renderer';
import InfoCard from 'pages/Home/InfoCard';
import { themed } from 'utils/tests';

describe('pages/Home/InfoCard', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      themed(<InfoCard />),
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
