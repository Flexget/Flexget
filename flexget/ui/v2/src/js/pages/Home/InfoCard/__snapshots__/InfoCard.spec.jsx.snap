// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/Home/InfoCard renders correctly 1`] = `
.glamor-3 {
  margin: 0 auto;
}

.glamor-0 {
  background-color: #ef6c00;
}

.glamor-1 {
  font-weight: bold;
}

.glamor-2 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

@media (min-width:600px) {
  .glamor-3 {
    width: 50%;
  }
}

<div
  className="MuiPaper-root-1 MuiPaper-shadow2-5 MuiPaper-rounded-2 glamor-3"
>
  <div
    className="MuiCardContent-root-33 MuiCardHeader-root-28 glamor-0"
  >
    <div
      className="MuiCardHeader-content-30"
    >
      <span
        className="MuiTypography-root-34 MuiTypography-headline-39 MuiCardHeader-title-31"
      >
        Flexget Web Interface
      </span>
      <span
        className="MuiTypography-root-34 MuiTypography-body1-43 MuiTypography-colorSecondary-55 MuiCardHeader-subheader-32"
      >
        Under Development
      </span>
    </div>
  </div>
  <div
    className="MuiCardContent-root-33"
  >
    <p
      className="glamor-1"
    >
      We need your help! If you are a React developer or can help with the layout/design/css then please join in the effort!
    </p>
    <p>
      The interface is not yet ready for end users. Consider this preview only state.
    </p>
    <p>
      If you still use it anyways, please report back to us on how well it works, issues, ideas etc...
    </p>
    <p>
      There is a functional API with documentation available at 
      <a
        href="/api"
      >
        /api
      </a>
    </p>
    <p>
      More information: 
      <a
        href="http://flexget.com/wiki/Web-UI/v2"
        rel="noopener noreferrer"
        target="_blank"
      >
        http://flexget.com/wiki/Web-UI/v2
      </a>
    </p>
    <p>
      Gitter Chat: 
      <a
        href="https://gitter.im/Flexget/Flexget"
        rel="noopener noreferrer"
        target="_blank"
      >
        https://gitter.im/Flexget/Flexget
      </a>
    </p>
  </div>
  <div
    className="MuiCardActions-root-58 glamor-2"
  >
    <a
      aria-label="Github"
      className="MuiButtonBase-root-69 MuiIconButton-root-60 MuiCardActions-actionSpacing-59"
      href="https://github.com/Flexget/Flexget"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      rel="noopener noreferrer"
      tabIndex={0}
      target="_blank"
    >
      <span
        className="MuiIconButton-label-66"
      >
        <span
          aria-hidden="true"
          className="material-icons MuiIcon-root-71 MuiIconButton-icon-67 fa fa-github"
        />
      </span>
      <span
        className="MuiTouchRipple-root-78"
      />
    </a>
    <a
      aria-label="Flexget.com"
      className="MuiButtonBase-root-69 MuiIconButton-root-60 MuiCardActions-actionSpacing-59"
      href="https://flexget.com"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      rel="noopener noreferrer"
      tabIndex={0}
      target="_blank"
    >
      <span
        className="MuiIconButton-label-66"
      >
        <span
          aria-hidden="true"
          className="material-icons MuiIcon-root-71 MuiIconButton-icon-67 fa fa-home"
        />
      </span>
      <span
        className="MuiTouchRipple-root-78"
      />
    </a>
    <a
      aria-label="Gitter"
      className="MuiButtonBase-root-69 MuiIconButton-root-60 MuiCardActions-actionSpacing-59"
      href="https://gitter.im/Flexget/Flexget"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      rel="noopener noreferrer"
      tabIndex={0}
      target="_blank"
    >
      <span
        className="MuiIconButton-label-66"
      >
        <span
          aria-hidden="true"
          className="material-icons MuiIcon-root-71 MuiIconButton-icon-67 fa fa-comment"
        />
      </span>
      <span
        className="MuiTouchRipple-root-78"
      />
    </a>
    <a
      aria-label="Forum"
      className="MuiButtonBase-root-69 MuiIconButton-root-60 MuiCardActions-actionSpacing-59"
      href="https://discuss.flexget.com"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      rel="noopener noreferrer"
      tabIndex={0}
      target="_blank"
    >
      <span
        className="MuiIconButton-label-66"
      >
        <span
          aria-hidden="true"
          className="material-icons MuiIcon-root-71 MuiIconButton-icon-67 fa fa-forumbee"
        />
      </span>
      <span
        className="MuiTouchRipple-root-78"
      />
    </a>
  </div>
</div>
`;
